src/stability_sdk/animation.py

```diff
@@ -366,100 +434,183 @@ def render(self) -> Generator[Image.Image, None, None]:
+        if not self.args.mask_invert:
+            self.mask = 255 - self.mask
+
+    def transform_2d(self, frame_idx) -> Optional[np.ndarray]:
+        if not len(self.prior_frames):
+            return None
+
+        args, frame_args = self.args, self.frame_args
+        angle = frame_args.angle_series[frame_idx]
+        scale = frame_args.zoom_series[frame_idx]
+        dx = frame_args.translation_x_series[frame_idx]
+        dy = frame_args.translation_y_series[frame_idx]
+
+        if angle == 0.0 and scale == 1.0 and dx == 0.0 and dy == 0.0:
+            return None
+
+        if args.accumulate_xforms:
+            frame_args = self.frame_args
+
+            # create xform for the current frame
+            xform = make_xform_2d(args.width, args.height, angle, scale, dx, dy)
+
+            # apply xform to prior frames running xforms
+            for i in range(len(self.prior_xforms)):
+                self.prior_xforms[i] = np.matmul(self.prior_xforms[i], xform)
+
+            # warp prior diffused frames by accumulated xforms
+            for i in range(len(self.prior_diffused)):
+                matrix = self.prior_xforms[i].copy().flatten().tolist()
+                op = generation.TransformOperation(
+                    warp2d=generation.TransformWarp2d(
+                        border_mode = border_mode_from_str_2d(args.border),
+                        matrix=generation.TransformMatrix(data=matrix)
+                ))                
+                xformed, _ = self.api.transform([self.prior_diffused[i]], [op])
+                self.prior_frames[i] = xformed[0]
+
+            return None
+        else:
+            op = warp2d_op(
+                border_mode=args.border,
+                rotate=angle, scale=scale, 
+                translate_x=dx, translate_y=dy, 
+            )
+            self.prior_frames, mask = self.api.transform(self.prior_frames, [op])
+            return mask
+
+    def transform_3d(self, frame_idx) -> Optional[np.ndarray]:
+        if not len(self.prior_frames):
+            return None
+
+        args, frame_args = self.args, self.frame_args
+        ops = [
+            depthcalc_op(blend_weight=args.midas_weight),
+            warp3d_op(
+                border_mode = args.border,
+                translate_x = frame_args.translation_x_series[frame_idx],
+                translate_y = frame_args.translation_y_series[frame_idx],
+                translate_z = frame_args.translation_z_series[frame_idx],
+                rotate_x = frame_args.rotation_x_series[frame_idx],
+                rotate_y = frame_args.rotation_y_series[frame_idx],
+                rotate_z = frame_args.rotation_z_series[frame_idx],
+                near_plane = args.near_plane,
+                far_plane = args.far_plane,
+                fov = frame_args.fov_series[frame_idx],
+            )
+        ]
+        self.prior_frames, mask = self.api.transform(self.prior_frames, ops)
```

Comment:
so rather than just estimating the warp for the current frame from the prior frames, we're continuously warping previous frames as long as they're in the two-frame context window? I can imagine that helps coherence, but probably complicates resume. Not saying it's a bad path forward, just let me know if I'm misunderstanding what's going on here.

Reply:
`prior_diffused` are the last results of Generate calls. `prior_frames` are those prior diffused frames forward warped to current camera space.  `prior_xforms` track the accumulated transform since last diffusion. the diffused frames are moved forward by their corresponding xforms then interpolated to get a frame between diffusion results.

Reply:
hmm... i'm not a huge fan of "prior_diffused" since the animation machinery here is generic to anything where you can do img2img, e.g. GANs. maybe we could change `prior_diffused` -> `prior_frames_raw` and `prior_frames` -> `prior_frames_after_xform`? or somesuch. not in love with the specific names I've suggested here, but I think we can make it clearer that `prior_frames` is specifically images that are ready to be used as initial state for a new generative process, whereas `prior_diffused` are the vanilla outputs from an earlier generative process.